src/agent/src/device.rs

```diff
@@ -436,217 +426,242 @@ fn scan_scsi_bus(scsi_addr: &str) -> Result<()> {
     Ok(())
 }
 
-// update_spec_device updates the device list in the OCI spec to make
-// it include details appropriate for the VM, instead of the host.  It
-// is given the host path to the device (to locate the device in the
-// original OCI spec) and the VM path which it uses to determine the
-// VM major/minor numbers, and the final path with which to present
-// the device in the (inner) container
-#[instrument]
-fn update_spec_device(
-    spec: &mut Spec,
-    devidx: &DevIndex,
-    host_path: &str,
-    vm_path: &str,
-    final_path: &str,
-) -> Result<()> {
-    let major_id: c_uint;
-    let minor_id: c_uint;
+#[derive(Debug, Clone)]
+struct SpecUpdate {
+    // the major and minor numbers for the device within the guest
+    guest_major: i64,
+    guest_minor: i64,
+    // an optional new path to update the device to in the "inner"
+    // container specification
+    final_path: Option<String>,
+    // optional corrections for PCI addresses associated with the
+    // device
+    pci: Vec<(pci::Address, pci::Address)>,
+}
+
+impl SpecUpdate {
+    fn from_vm_path<T: AsRef<Path>>(vm_path: T) -> Result<Self> {
+        let vm_path = vm_path.as_ref();
 
-    // If no container_path is provided, we won't be able to match and
-    // update the device in the OCI spec device list. This is an error.
-    if host_path.is_empty() {
-        return Err(anyhow!("Host path cannot empty for device"));
+        if !vm_path.exists() {
+            return Err(anyhow!(
+                "VM device path {} doesn't exist",
+                vm_path.display()
+            ));
+        }
+
+        let devid = fs::metadata(vm_path)?.rdev();
+        let guest_major = stat::major(devid);
+        let guest_minor = stat::minor(devid);
+
+        Ok(SpecUpdate {
+            guest_major: guest_major as i64,
+            guest_minor: guest_minor as i64,
```

Comment:
You could simplify this:

```suggestion
        let guest_major = stat::major(devid) as i64;
        let guest_minor = stat::minor(devid) as i64;

        Ok(SpecUpdate {
            guest_major,
            guest_minor,
```

Reply:
Updated, thanks.